src/ExtraValuesAssigner.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+namespace Wikibase\DataModel;
+
+/**
+ * Assigns extra values to the serialization of an object
+ * based on the given extra values providers.
+ *
+ * @since 2.1
+ *
+ * @licence GNU GPL v2+
+ * @author Bene* < benestar.wikimedia@gmail.com >
+ */
+class ExtraValuesAssigner {
+
+	/**
+	 * @var ExtraValuesProvider[]
+	 */
+	private $extraValuesProviders;
+
+	/**
+	 * @param ExtraValuesProvider[] $extraValuesProviders
+	 */
+	public function __construct( array $extraValuesProviders = array() ) {
+		$this->extraValuesProviders = $extraValuesProviders;
+	}
+
+	/**
+	 * @param array $serialization
+	 * @param mixed $object
+	 * @return array
+	 */
+	public function addExtraValues( $serialization, $object ) {
+		foreach ( $this->extraValuesProviders as $extraValuesProvider ) {
+			if ( $extraValuesProvider->isProviderFor( $object ) ) {
+				$serialization = array_merge(
+					$extraValuesProvider->getExtraValues( $object ),
+					$serialization
```

Comment:
> > Having a solution like this will make it easier to understand.

 > How?

Well, do you currently understand what is happening in the ResultBuilder in Repo? :P This would make it 1000 times clearer.

 > > It would also increase the efficiency of our serialization code when adding 'all of the things' we currently do in the API.

 > Can you elaborate?

Currently to add all of the other stuff in that we need through the api for example we do lots of looping through the serialization, as we cant add things to it when it is initially created.
I am guessing that this approach would be more efficient, but I am fairly certain loosing the loops would help.

 > > Having it in a separate component would make it version-able.

 > True, though you can just put the current code into its own component as well no?

That is probably fairly true, although the current stuff binds to mediawiki things, although this could probably be avoided, although it would probably involve something very similar to the approach in this PR, but perhaps still with lots of looping (if we don't like this PR)